astroplan/core.py

```diff
@@ -162,56 +164,336 @@ def altaz(self, time, target=None, obswl=None):
+
+    @u.quantity_input(horizon=u.deg)
+    def _two_point_interp(self, times, altitudes, horizon=0*u.deg):
+        '''
+        Do linear interpolation between two `altitudes` at
+        two `times` to determine the time where the altitude
+        goes through zero.
+
+        Parameters
+        ----------
+        times : `~astropy.time.Time`
+            Two times for linear interpolation between
+
+        altitudes : array of `~astropy.units.Quantity`
+            Two altitudes for linear interpolation between
+
+        horizon : `~astropy.units.Quantity`
+            Solve for the time when the altitude is equal to
+            reference_alt.
+
+        Returns
+        -------
+        t : `~astropy.time.Time`
+            Time when target crosses the horizon
+
+        '''
+        slope = (altitudes[1] - altitudes[0])/(times[1].jd - times[0].jd)
+        return Time(times[1].jd - ((altitudes[1] - horizon)/slope).value,
+                    format='jd')
+
+    def _calc_riseset(self, time, target, prev_next, rise_set, horizon, N=150):
+        '''
+        Time at next rise/set of `target`.
 
         Parameters
         ----------
         time : `~astropy.time.Time`
-        """
-        raise NotImplementedError()
+            Time of observation
 
-    def midnight(self, time):
-        """
-        Returns the local, solar midnight time.
+        target : `~astropy.coordinates.SkyCoord`
+            Position of target or multiple positions of that target
+            at multiple times (if target moves, like the Sun)
+
+        prev_next : str - either 'previous' or 'next'
+            Test next rise/set or previous rise/set
+
+        rise_set : str - either 'rising' or 'setting'
+            Compute prev/next rise or prev/next set
+
+        location : `~astropy.coordinates.EarthLocation`
+            Location of observer
+
+        horizon : `~astropy.units.Quantity`
+            Degrees above/below actual horizon to use
+            for calculating rise/set times (i.e.,
+            -6 deg horizon = civil twilight, etc.)
+
+        N : int
+            Number of altitudes to compute when searching for
+            rise or set.
+
+        Returns
+        -------
+        ret1 : `~astropy.time.Time`
+            Time of rise/set
+        '''
+        if prev_next == 'next':
+            times = time + np.linspace(0, 1, N)*u.day
+        else:
+            times = time + np.linspace(-1, 0, N)*u.day
+
+        altitudes = self.altaz(times, target).alt
+        horizon_crossing_limits = self._horiz_cross(times, altitudes, rise_set,
+                                                    horizon,
+                                                    return_limits=True,
+                                                    return_alts=True)
+        return self._two_point_interp(*horizon_crossing_limits, horizon=horizon)
+
+    @u.quantity_input(horizon=u.deg)
+    def calc_rise(self, time, target, which='nearest', horizon=0*u.degree):
```

Comment:
Usually in Python one should never have a mutable object as default value for keyword arguments.
@eteq, @adrn – Is this considered the best solution for functions that take quantities as input?